### ISSUE TITLE
Recommended way to enforce option selection for sub-subprojects?

### ISSUE THREAD

**Issue by phillipjohnston** [mesonbuild/meson#7884]
I'm working on a framework that uses Meson as the build system. I have a core project, and that project pulls in a number of dependencies and incorporates them into the build using `subproject`. The framework needs to enforce some configuration options on the dependencies, while providing other options that the user can adjust.

However, I can't find a good way to actually enforce the defaults that I need when I include the core project as a subproject. When I build the dependencies, the default options specified in the dependency projects' meson_options.txt file get used. It seems like `project(default_options)` are ignored if I'm a subproject, and it seems that `subproject(default_options:)` doesn't actually work (#6728, if it's still valid - I couldn't get it working at least). 

The only solution I've found so far is to copy the options from each dependency into the application that is built with the framework. But this makes the option file quite confusing, because now we're listing options for sub-sub-projects - and that's confusing because some of those options should not be modified at all if the build is going to succeed.

Is there an approach to managing this that I'm missing? Or is Meson just not a good choice for this type of build structure?

Edit: Just to be clear, I'm not looking for a way to hide options, as much as for a way to provide defaults that are set in the core project and not required to be listed in the application's meson_options.txt file.

**Comment by phillipjohnston** [comment]
I just found #7854, which seems like this should actually work... but I must be missing something, because this was definitely not working for me when I tried it:

> It seems like project(default_options) are ignored if I'm a subproject

**Comment by Akaricchi** [comment]
> It seems like project(default_options) are ignored if I'm a subproject

Yeah, that seems to be by design. Whether this behavior is desirable is up for debate... cc @dcbaker @xclaesse 

I think currently you're expected to set all of the defaults in the main project's `project()` statement.

**Comment by dcbaker** [comment]
It is by design, allowing subprojects to alter other subprojects can introduce ask kinds of bugs and race conditions. With meson 0.56 you'll be able to put all of the options in a cross or native file, which may be an effective way to handle it.

**Comment by phillipjohnston** [comment]
I have main project which includes library A, which configures and includes library B according to its needs. Main project doesn’t know anything about library B, because it’s used by library A to satisfy its requirements. 

Why does main project need to set the options for library B, which it otherwise knows nothing about? That’s where the bug was introduced in my program.

**Comment by phillipjohnston** [comment]
After sleeping on this, one way I can think to solve this problem is to transform options into additional build targets and dependencies. This works for some options, with the tradeoff of increasing the number of targets/deps in the build rules, since I have a variant of each target for the specific options.

**Comment by bonzini** [comment]
Closing as duplicate since the Meson-specific discussion happened there.
